From c6b60f3480273cd7c3182795abd6cfcf07fa078a Mon Sep 17 00:00:00 2001
Message-Id: <c6b60f3480273cd7c3182795abd6cfcf07fa078a.1487774148.git.anders.wallin@windriver.com>
In-Reply-To: <0f51b4bf70b312f58d1f7eeebd3af3084511e1b9.1487774148.git.anders.wallin@windriver.com>
References: <0f51b4bf70b312f58d1f7eeebd3af3084511e1b9.1487774148.git.anders.wallin@windriver.com>
From: Liguang Li <liguang.li@windriver.com>
Date: Thu, 15 Dec 2016 18:56:20 +0800
Subject: [PATCH 2/3] Add ctx_len field to ring buffer context

Issue : LINCCM-995

Allow extending the structure beyond its original size.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Signed-off-by: Liguang Li <liguang.li@windriver.com>
Signed-off-by: Anders Wallin <anders.wallin@windriver.com>
---
 include/lttng/ringbuffer-config.h | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/include/lttng/ringbuffer-config.h b/include/lttng/ringbuffer-config.h
index 5ae0d8a6..c95d35cb 100644
--- a/include/lttng/ringbuffer-config.h
+++ b/include/lttng/ringbuffer-config.h
@@ -256,10 +256,21 @@ struct lttng_ust_lib_ring_buffer_ctx {
 					 */
 	uint64_t tsc;			/* time-stamp counter value */
 	unsigned int rflags;		/* reservation flags */
-	unsigned int padding1;		/* padding to realign on pointer */
+	/*
+	 * The field ctx_len is the length of struct
+	 * lttng_ust_lib_ring_buffer_ctx as known by the user of
+	 * lib_ring_buffer_ctx_init.
+	 */
+	unsigned int ctx_len;
 	void *ip;			/* caller ip address */
 	void *priv2;			/* 2nd priv data */
 	char padding2[LTTNG_UST_RING_BUFFER_CTX_PADDING];
+	/*
+	 * This is the end of the initial fields expected by the original ABI
+	 * between probes and UST. Only the fields above can be used if
+	 * ctx_len is 0. Use the value of ctx_len to find out which of the
+	 * following fields may be used.
+	 */
 };
 
 /**
@@ -291,7 +302,7 @@ void lib_ring_buffer_ctx_init(struct lttng_ust_lib_ring_buffer_ctx *ctx,
 	ctx->cpu = cpu;
 	ctx->rflags = 0;
 	ctx->handle = handle;
-	ctx->padding1 = 0;
+	ctx->ctx_len = sizeof(struct lttng_ust_lib_ring_buffer_ctx);
 	ctx->ip = 0;
 	ctx->priv2 = priv2;
 	memset(ctx->padding2, 0, LTTNG_UST_RING_BUFFER_CTX_PADDING);
-- 
2.11.0

